src/ramdb.cpp

```diff
@@ -0,0 +1,385 @@
+    auto file                 = std::ofstream{time_file_path};
+
+    if(!file)
+    {
+        MIOPEN_LOG_E("Cannot update database modification time: " + time_file_path);
+        return;
+    }
+
+    file << time.count();
+}
+
+#define MIOPEN_VALIDATE_LOCK(lock)                       \
+    do                                                   \
+    {                                                    \
+        if(!(lock))                                      \
+            MIOPEN_THROW("Db lock has failed to lock."); \
+    } while(false)
+
+static std::chrono::seconds GetLockTimeout() { return std::chrono::seconds{60}; }
+
+using exclusive_lock = std::unique_lock<LockFile>;
+
+RamDb::RamDb(std::string path, bool is_system) : PlainTextDb(path, is_system) {}
+
+RamDb& RamDb::GetCached(const std::string& path, bool is_system)
+{
+    // NOLINTNEXTLINE (cppcoreguidelines-avoid-non-const-global-variables)
+    static std::mutex mutex;
+    const std::lock_guard<std::mutex> lock{mutex};
+
+    // NOLINTNEXTLINE (cppcoreguidelines-avoid-non-const-global-variables)
+    static auto instances = std::map<std::string, RamDb*>{};
+    const auto it         = instances.find(path);
+
+    if(it != instances.end())
+        return *it->second;
+
+    // The ReadonlyRamDb objects allocated here by "new" shall be alive during
+    // the calling app lifetime. Size of each is very small, and there couldn't
+    // be many of them (max number is number of _different_ GPU board installed
+    // in the user's system, which is _one_ for now). Therefore the total
+    // footprint in heap is very small. That is why we can omit deletion of
+    // these objects thus avoiding bothering with MP/MT syncronization.
+    // These will be destroyed altogether with heap.
+    auto instance = new RamDb{path, is_system};
+    instances.emplace(path, instance);
+    if(!DisableUserDbFileIO)
+    {
+        const auto prefetch_lock = exclusive_lock(instance->GetLockFile(), GetLockTimeout());
+        MIOPEN_VALIDATE_LOCK(prefetch_lock);
+        instance->Prefetch();
+    }
+    return *instance;
+}
+
+boost::optional<DbRecord> RamDb::FindRecord(const std::string& problem)
+{
+    const auto lock = exclusive_lock(GetLockFile(), GetLockTimeout());
+    MIOPEN_VALIDATE_LOCK(lock);
+
+    if(!ValidateUnsafe())
+    {
+        MIOPEN_LOG_I2("RamDb file is newer than cache, prefetching");
+        Prefetch();
+    }
+
+    return FindRecordUnsafe(problem);
+}
+
+bool RamDb::StoreRecord(const DbRecord& record)
+{
+    const auto& key = record.GetKey();
+    MIOPEN_LOG_I2("Trying to store record at key " << key << " in cache for file "
+                                                   << GetFileName());
+    const auto lock = exclusive_lock(GetLockFile(), GetLockTimeout());
+    MIOPEN_VALIDATE_LOCK(lock);
+
+    if(!DisableUserDbFileIO)
+    {
+        if(!StoreRecordUnsafe(record))
+            return false;
+        UpdateDbModificationTime(GetFileName());
```

Comment:
This `UpdateDbModificationTime` function will update the modification time of user_find_db. Thus `time` > `file_read_time` in the following function `UpdateCacheEntryUnsafe`, which result in `is_valid` being false, then, the code block in the if statement is not executed and `file_read_time` not updated. The next time finding needs to prefetch() again. 
```
const auto is_valid = ValidateUnsafe();

    if(!DisableUserDbFileIO)
        UpdateDbModificationTime(GetFileName());

    if(is_valid)
    {
        const auto& key = record.GetKey();
        const auto it   = cache.find(key);
        auto ss         = std::ostringstream{};
        record.WriteIdsAndValues(ss);

        if(it != cache.end())
        {
            auto& item   = it->second;
            item.content = ss.str();
        }
        else
        {
            cache.emplace(key, CacheItem{-1, ss.str()});
        }
        file_read_time = ramdb_clock::now();
    }
```
But what you want is not to prefetch again if there is no other thread to modify the `user_find_db`, right?

Reply:
@atamazov and @DrizztDoUrden , could you take a look at the above comments?